### PR TITLE
feat: use gossiped commit in sync

### LINKED ISSUE #484: Keep full nodes in-sync with aggregator

### PR DESCRIPTION
**Pull request by tzdybal**
Changes:
* gossip commits with headers (using SignedHeader)
* use gossiped commit during sync

Resolves #484.

### CODE REVIEW

**Review by adlerjohn — DISMISSED** [review]
utACK

File: node/node.go [hunk 1]
```diff
@@ -325,7 +334,7 @@ func (n *Node) newHeaderValidator() p2p.GossipValidator {
 	}
 }
 
-// newCommitValidator returns a pubsub validator that runs basic checks and forwards
+// newCommitValidator returns a pubsub vali
```
Comment (Wondertan) [inline node/node.go:337]: Seems like an accidental remove

File: block/manager.go [hunk 2]
```diff
@@ -218,49 +225,81 @@ func (m *Manager) SyncLoop(ctx context.Context) {
...
+
+	return nil
+}
+
+func (m *Manager) getLastCommit() *types.Commit {
+	ptr := m.lastCommit.Load()
+	if ptr != nil {
+		return m.lastCommit.Load().(*types.Commit)
+	}
+	return nil
```
Comment (Wondertan) [inline block/manager.go:300]: readability nit:
```suggestion
	if ptr == nil {
		return nil
	}
	return return m.lastCommit.Load().(*types.Commit)
```

File: block/manager.go [hunk 3]
```diff
@@ -218,49 +225,81 @@ func (m *Manager) SyncLoop(ctx context.Context) {
...
+			if err != nil {
+				m.logger.Info("failed to sync next block", "error", err)
 			}
 		case <-ctx.Done():
 			return
 		}
 	}
 }
 
+func (m *Manager) trySyncNextBlock(ctx context.Context, daHeight uint64) error {
```
Comment (Wondertan) [inline block/manager.go:239]: It's kinda hard to grasp what happens in this method. A brief comment with an explanation would help.

Comment (Wondertan) [reply block/manager.go:239]: Even if it just an extraction of existing logic

File: block/manager.go [hunk 4]
```diff
@@ -205,9 +205,16 @@ func (m *Manager) SyncLoop(ctx context.Context) {
 				atomic.StoreUint64(&m.syncTarget, newHeight)
 				m.retrieveCond.Signal()
 			}
+			m.CommitInCh <- &header.Commit
 		case commit := <-m.CommitInCh:
 			// TODO(tzdybal): check if it's from right aggregator
-			m.lastCommit = commit
+			m.lastCommit.Store(commit)
+			err := m.trySyncNextBlock(ctx, 0)
```
Comment (Wondertan) [inline block/manager.go:212]: Why zero though? the `syncCache` for it will be empty, right?

File: node/node.go [hunk 5]
```diff
@@ -200,6 +200,15 @@ func (n *Node) headerPublishLoop(ctx context.Context) {
 			if err != nil {
 				n.Logger.Error("failed to gossip block header", "error", err)
 			}
+		case commit := <-n.blockManager.CommitOutCh:
+			commitBytes, err := commit.MarshalBinary()
+			if err != nil {
+				n.Logger.Error("failed to serialize block commit", "error", err)
+			}
+			err = n.P2P.GossipCommit(ctx, commitBytes)
```
Comment (Wondertan) [inline node/node.go:208]: Feels like we should `continue` in case of an error

File: node/node.go [hunk 6]
```diff
@@ -200,6 +200,15 @@ func (n *Node) headerPublishLoop(ctx context.Context) {
...
+			if err != nil {
+				n.Logger.Error("failed to gossip block commit", "error", err)
+			}
```
Comment (Wondertan) [inline node/node.go:211]: Btw, it's not necessary to make sync loop only responsible for publishing/gossiping messages, and the code that needs to publish can use GossipCommit directly. But I think you understand this and made `CommitOutCh` for symmetry with `CommitInCh` 😁  which is fine. However, it's not the best justification for bottlenecking both msg publishing and the syncloop.

File: block/manager.go [hunk 7]
```diff
@@ -482,13 +519,8 @@ func (m *Manager) exponentialBackoff(backoff time.Duration) time.Duration {
...
-func (m *Manager) publishHeader(block *types.Block) {
-	m.HeaderOutCh <- &block.Header
-}
-
-// TODO(tzdybal): consider inlining
-func (m *Manager) publishCommit(commit *types.Commit) {
-	m.CommitOutCh <- commit
+func (m *Manager) publishSignedHeader(block *types.Block, commit *types.Commit) {
+	m.HeaderOutCh <- &types.SignedHeader{Header: block.Header, Commit: *commit}
 }
```
Comment (Wondertan) [inline block/manager.go:524]: If we now combine publishing with one HeaderOutCh, do we use CommitOutCh elsewhere?

File: block/manager.go (hunk 4 above)
Comment (tzdybal) [reply block/manager.go:212]: Because we don't know DA height here, as commit was gossiped via P2P.

File: block/manager.go (hunk 7 above)
Comment (tzdybal) [reply block/manager.go:524]: Good catch! I missed this while refactoring.

File: node/node.go (hunk 6 above)
Comment (tzdybal) [reply node/node.go:211]: Good catch! I missed this bit during refactoring :facepalm:

File: block/manager.go [hunk 8]
```diff
@@ -277,7 +281,7 @@ func (m *Manager) trySyncNextBlock(ctx context.Context, daHeight uint64) error {
 			return fmt.Errorf("failed to save block responses: %w", err)
 		}
 
-		if m.daHeight > newState.DAHeight {
```
Comment (Wondertan) [inline block/manager.go]: Was that a bug you accidentally caught? 😁

File: block/manager.go [hunk 9]
```diff
@@ -235,6 +234,11 @@ func (m *Manager) SyncLoop(ctx context.Context) {
 	}
 }
 
+// trySyncNextBlock tries to progress one step (one block) in sync process.
+//
+// To be able to apply block and height h, we need to have its Commit. It is contained in block at height h+1.
+// If block at height h+1 is not available, value of last gossiped commit is checked.
+// If commit for block h is available, we proceed with sync process, and remove synced block from sync cache.
```
Comment (Wondertan) [inline block/manager.go:241]: 🙏🏻

**Review by Wondertan — APPROVED** [review]
LGTM

This all is very interesting. I will likely visit the Optimint repo much more often now!

File: block/manager.go (hunk 8 above)
Comment (tzdybal) [reply block/manager.go]: Yes!

**Pull request merged**